router/oauth2/token_endpoint.go

```diff
@@ -119,23 +168,11 @@ func (h *Handler) tokenEndpointAuthorizationCodeHandler(c echo.Context) error {
 	}
 
 	// トークン発行
-	newToken, err := h.Repo.IssueToken(client, code.UserID, client.RedirectURI, code.Scopes, h.AccessTokenExp, h.IsRefreshEnabled)
+	res, err := h.issueToken(client, code.UserID, code.Scopes, code.OriginalScopes, true)
```

Comment:
>offline_access
OPTIONAL. OAuth 2.0 Refresh Token の発行を要求する. Request Token は End-User の関与を必要とせずに End-User の UserInfo Endpoint にアクセスするための Access Token を取得するために利用される.

> OP は Refresh Token を発行する際は必ず明示的な同意を得なければならない (MUST)

個のトークンを用いてUserInfo Endpointにアクセス可能にするのであれば、ユーザーの明示的な同意なく発行してはいけないです。

Reply:
https://openid-foundation-japan.github.io/openid-connect-core-1_0.ja.html#OfflineAccess

Reply:
見てくれてありがとうございます。ちょっと気になったので調べてみました。

結論からいうと、推奨はされるもの**かもしれない**が、絶対(must)ではなく、そんなに問題ではないと思います。

元の文章の

> The OP MUST always obtain consent to returning a Refresh Token that enables **offline access** to the requested resources.
https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess

は、**offline accessを可能にする**refresh tokenについて述べられていそうです。この段落に `When offline access is requested,` がかかっているとも読めそうです。

その下にも書いてある通り、

> The use of Refresh Tokens is **not exclusive to the offline_access use case**. The Authorization Server MAY grant Refresh Tokens in other contexts that are beyond the scope of this specification.
https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess

offline accessを要求し、かつconsentしないとrefresh tokenを発行してはいけない、なんてことはないよ、とあります。

もしそうならOAuth 2.0 (https://datatracker.ietf.org/doc/html/rfc6749#section-6) では基本的に発行が (implicit flowなどを除いて) optionalと書かれているrefresh tokenの記述と矛盾するよな、と思って調べてました。

https://github.com/manfredsteyer/angular-oauth2-oidc/issues/1241 にも似た議論があります。
https://oauth.jp/blog/2012/10/01/oauthofflineaccess/, https://creators-note.chatwork.com/entry/2018/08/07/102455, https://stackoverflow.com/questions/30637984/what-does-offline-access-in-oauth-mean も参考になりました。

そして最後に"offline access"の定義なんですが、ちょっと難しいと思います。

> Note that differentiating between online and offline access from the server can be difficult especially for native clients.
https://openid-foundation-japan.github.io/openid-connect-core-1_0.ja.html#OfflineAccessPrivacy

ともある通り、本質的に"online access"と"offline access"を区別するのは難しいようです。
Access tokenの有効期間が1時間とかで、consent+offline accessを要求しないとrefresh tokenをくれないGoogleの実装が一番セキュア(?)かもしれないですが、他のOAuth2/OIDCサーバーの実装は普通にrefresh tokenをくれたりするそうです(traQ含め)。
つまり、ユーザーがrevokeするまで"offline access"もできますし、変に複雑なよりはまあそれで十分便利なんじゃないでしょうか。

と言いましたが、traQではデフォルトでrefresh token発行できない設定で、access tokenの有効期間は1年(ちょっと長い?)、refresh tokenの有効期間は無しになっています。
本番でもrefresh tokenは発行できないようになってるはずです。

Reply:
なので、https://openid-foundation-japan.github.io/openid-connect-core-1_0.ja.html#OfflineAccess の訳が結構ガバガバというのもありそうです。("Request Token"ってなんだよ...)

Reply:
返信ありがとうございます。

OAuth 2.0 の文脈では確かにその通りで、Refresh tokenを発行するのにOffline Accessは必要ではないです。

しかし、OIDCの文脈になると話が変わって必要になるというのが僕の認識です。
（とはいえ、上に張ってあるリンクのとおり、現実の実装としてどうなってるかというとそこまで対応されてないことも多い気がするので対応するかどうかの判断は任せます。）

そもそもOAuth 2.0 は「認可」の仕様であり、あくまで「ユーザー認証」は副作用です。（そのためにOIDCができた）

そのため、OAuth「ユーザー認証」を必要としない Access Token / Refresh Token の発行手段が存在し、具体的にはクライアント認証によって認可を行う「Client Credentials Flow」がそれに該当します。（実際OIDCにはこのフローが存在しない）

OIDC は OAuthの拡張である以上、authorization codeなど、共通のトークンが発行されることがあるわけですが、OIDC ではその二つを明確に区別していて、たとえば Token Endpoint における authorization code の検証では、
> The Authorization Server MUST validate the Token Request as follows:
> ~~~~ 
> Verify that the Authorization Code used was issued in response to an OpenID Connect Authentication Request (so that an ID Token will be returned from the Token Endpoint).

https://openid.net/specs/openid-connect-core-1_0.html#TokenRequestValidation

と、ユーザー認証が行われたフローで発行された authorization code であることを検証するフェーズがあります。(MUST)

当然 Access Token / Refresh Token もその二つの区別はしないといけなくて、UserInfo Endpointではそのアクセストークンを認証したユーザーのUserInfoが返却されるわけですが、Client Credentials Flowではユーザー認証が行われないため、判別することが原理的に不可能です。（追加のクエリパラメータで渡してどうにかする方法はありますが、実装依存なので割愛）

という風に、二つの Token を区別したとき、offline access は、
> This scope value requests that an OAuth 2.0 Refresh Token be issued that can be used to obtain an Access Token that grants access to the End-User's UserInfo Endpoint even when the End-User is not present (not logged in)

となっていることから、ユーザー認証が行われるOIDC の Access Token についてのみの制約であると解釈できます。（UserInfo Endpointをたたけるのはユーザー認証済みのAccess Tokenしかないため）

（ちょっとこれからMTGなので一時保存）（続きます）

Reply:
なるほど、OAuth 2.0の範囲で発行されたAccess Token/Refresh Tokenなのか、OIDCの定義の範囲で発行されたAccess Token/Refresh Tokenなのかを区別して見ると、提示してくれた文章は後者 (UserInfo Endpointを許可するAccess Token/Refresh Token) について書かれた制約ということですね。

そうなると、OIDCの範囲の (`openid` `profile` scopeを含む) Token Requestの場合 (かつ `offline_access` でない場合)、そこで発行するRefresh TokenにはUserInfo Endpointが読めるAccess Tokenの発行を許可しない、という実装になるんでしょうか。
ただ、それをする場合、今のAccess Tokenのデフォルト有効期間をせめて1年より短くするなどしないと意味がなさそうなので、元のtraQのOAuth 2.0の実装にも触れることになるかなと思います。

今のtraQの持つ個人情報の多さと、便利さを天秤にかけて、そこまで細かくは触りたくはないかな...とも正直思っています。（最終的には実装の多さにもなるので、メンテナー @logica も判断してほしいです。）
traPortalはもちろん個人情報を多く持つので、有効期間を短くするなど、ちゃんとsecurity measureは最低限必要だと思いますが。（そのためにもscratch実装は辛いのでは...）

Reply:
> そうなると、OIDCの範囲の (openid profile scopeを含む) Token Requestの場合 (かつ offline_access でない場合)、そこで発行するRefresh TokenにはUserInfo Endpointが読めるAccess Tokenの発行を許可しない、という実装になるんでしょうか。

これtokiさんが挙げてくれた "offline"の定義の難しさにもつながると思うんですが、たぶん（認証フローで得た）Access Tokenが有効な期間は"online"と解釈してよいのではないかと思います。

Refresh Tokenを使うと、ユーザーの明示的な操作なしにAccess Tokenが取得できるので"offline"と解釈できるのかなと思います。

したがって、Access Tokenにofflineのスコープがあるかどうかで判別する必要はないと考えています（UserInfo Endpointの節にもその説明はないですし）



そもそもAccess Tokenって短い期間でExpireするものなので・・・（というか今一年も有効期間あるのか）

Reply:
あ、これRefresh Tokenの話か、であればそもそも"Refresh Tokenの発行を許可しない"が一番楽な気がします（そこら辺のScope管理めんどくさいので）

Reply:
とはいえそれもめんどくさいのは確かなので、最初にも言ったとおり、無くてもいいのかなとは思っています

> （とはいえ、上に張ってあるリンクのとおり、現実の実装としてどうなってるかというとそこまで対応されてないことも多い気がするので対応するかどうかの判断は任せます。）

Reply:
そうですね、「scopeにopenidが含まれる場合、refresh tokenの発行をしない」が一番実装も単純で、制約も満たせるので良い気がします

まあそもそも本番でも、一切refresh tokenは発行してないのですが